distributed/scheduler.py

```diff
@@ -7477,7 +7509,31 @@ def decide_worker(
         for ws in candidates:
             break
     else:
+        group: TaskGroup = ts._group
+        ws = group._last_worker
+
+        if valid_workers is not None:
+            total_nthreads = sum(wws._nthreads for wws in candidates)
+
+        group_tasks_per_worker = len(group) / total_nthreads
+
+        # Try to schedule sibling root-like tasks on the same workers, so subsequent reduction tasks
+        # don't require data transfer. Assumes `decide_worker` is being called in priority order.
+        if (
+            ws is not None  # there is a previous worker
+            and group._last_worker_tasks_left > 0  # previous worker not fully assigned
+            and group_tasks_per_worker > 1  # group is larger than cluster
+            and (  # is a root-like task (task group depends on very few tasks)
+                sum(map(len, group._dependencies)) < 5  # TODO what number
+            )
```

Comment:
Does this mean that only groups of size `N (=5)` are grouped together and everything above is distributed _randomly_?

If my understanding here is correct, we might want to choose 32 (+/-1?) since this is the default branching factor for tree reductions in dask, last time I checked

Reply:
No, this is restricing this behavior to task groups where the size of the task group is large, but the number of dependencies of that group is small.  This is true in situations like the following:

```python
df = dd.read_parquet(...)  # thousands of tasks, but zero dependenies

x = da.from_zarr(...)  # thousands of tasks, but one dependency (the zarr file)
```

But not in situations like the following:

```python
y = x + 1  # thousands of tasks, but thousands of dependencies
```